### PR TITLE
Fix writing non contiguous data to a compressed HDU.

### LINKED ISSUE #2150: Pyfits: Incorrect writing of non contiguous data in compressed FITS
Using astropy 0.3.dev4315, the following code prints False instead of True. If ImageHDU is substituted to CompImageHDU, it works correctly and prints True.

```
orig = np.arange(100, dtype=float).reshape((10, 10), order='f')
primary = fits.PrimaryHDU()
hdu = fits.CompImageHDU(orig)
hdulist = fits.HDUList([primary, hdu])
hdulist.writeto('test.fits', clobber=True)
actual = fits.open('test.fits')[1].data
print np.all(orig == actual)
```

### LINKED ISSUE #9955: Invalid cutout/slice data for rice tile compressed images.
EDIT: See https://github.com/astropy/astropy/issues/9955#issuecomment-587342185 for workaround.

### Description

When slicing rice tile compressed image, the output data are invalid. In some cases this gets magically fixed if `scale_back=True` flag is passed when opening the fits file, but on some other cases it has no effect.

To clarify what I mean by valid and invalid data, see attached examples. 

### Expected behavior

This is what I would expect to get:
valid

### Actual behavior

And this is what is produced:
invalid

(obviously the data are totally wrong, but the images show this much better than tables)

### Steps to Reproduce

This can be easily reproduced by:

```python
        outfile = "compressed_cut.fz"
        infile = "ADP.2020-02-12T10_26_25.400.fits.fz" # you can get the input file from http://dataportal.eso.org/dataPortal/file/ADP.2020-02-12T10:26:25.400
        hdus = fits.open(infile)
        slices = tuple([slice(start, stop) for (start, stop) in [(0, 4090), (0, 2040)]])
        hdus[1].data = hdus[1].data[slices]
        new_hdus = hdus[0:2]
        new_hdus.writeto(outfile, overwrite=True)
        hdus.close()
```

It's also pretty easy to see that the actual data saved to the file, are not what you'd expect. If we extend the above example to:

```python
        outfile = "compressed_cut.fz"
        infile = "ADP.2020-02-12T10_26_25.400.fits.fz"  # you can get the input file from http://dataportal.eso.org/dataPortal/file/ADP.2020-02-12T10:26:25.400
        hdus = fits.open(infile)
        slices = tuple([slice(start, stop) for (start, stop) in [(0, 4090), (0, 2040)]])
        hdus[1].data = hdus[1].data[slices]
        new_hdus = hdus[0:2]
        print(hdus[1].data[1][0:100]) # data we're presumably saving to the file
        new_hdus.writeto(outfile, overwrite=True)
        hdus.close()

        hdus = fits.open(outfile)
        print(hdus[1].data[1][0:100]) # data we're reading back from file
```

We get two very different sets of data printed out, even though they should be the same.

Interestingly enough, if I cast the data to `uint32`, then the output is correct:

```python
        infile = "ADP.2020-02-12T10_26_25.400.fits.fz" # you can get the input file from http://dataportal.eso.org/dataPortal/file/ADP.2020-02-12T10:26:25.400
        hdus = fits.open(infile)
        slices = tuple([slice(start, stop) for (start, stop) in [(0, 4090), (0, 2040)]])
        import numpy as np
        hdus[1].data = np.uint32(hdus[1].data[slices])
        new_hdus = hdus[0:2]
        new_hdus.writeto(outfile, overwrite=True)
        hdus.close()
```

But this in turn sets BSCALE to 1, and since the cast is to unsigned values, if there were negative values in the data, then they get turned into maxint, which is also a bit inconvenient.
I tried using the `do_not_scale_image_data=True` and  `scale_back=True` flags but they have no effect.

Am I doing something wrong here?

### System Details

Windows-10-10.0.17763-SP0
Python 3.7.3 (default, Apr 24 2019, 15:29:51) [MSC v.1915 64 bit (AMD64)]
Numpy 1.16.4
astropy 3.2.1

### PR DESCRIPTION
**Pull request by saimn**
Fixes #2150 and fixes #9955.

### CODE REVIEW

File: astropy/io/fits/tests/test_image.py [hunk 1]
```diff
@@ -1790,6 +1790,31 @@ def test_compressed_integers(self, dtype):
...
+        actual = fits.getdata(self.temp('test.fits'))
+        assert_equal(orig, actual)
+
+    def test_slice_and_write_comp_hdu(self):
+        """
+        Regression test for https://github.com/astropy/astropy/issues/9955
+        """
+        with fits.open(self.data('comp.fits')) as hdul:
+            hdul[1].data = hdul[1].data[:200, :100]
+            hdul[1].writeto(self.temp('test.fits'))
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_image.py]: Could you assert here that it's non-contiguous before?

File: CHANGES.rst [hunk 2]
```diff
@@ -357,6 +357,8 @@ astropy.io.fits
 - Fix ``fitsdiff`` when its arguments are directories that contain other
   directories. [#9711]
 
+- Fix writing non contiguous data to a compressed HDU. [#9958]
```
Comment (pllim) [inline CHANGES.rst]:
```suggestion
- Fix writing noncontiguous data to a compressed HDU. [#9958]
```

**Comment by bsipocz** [comment]
Always great to see when a super old issue is being fixed :) Thanks @saimn!

File: astropy/io/fits/tests/test_image.py (hunk 1 above)
Comment (saimn) [reply astropy/io/fits/tests/test_image.py]: Done.

File: CHANGES.rst (hunk 2 above)
Comment (saimn) [reply CHANGES.rst]: Done.

File: astropy/io/fits/tests/test_image.py [hunk 3]
```diff
@@ -1790,6 +1790,33 @@ def test_compressed_integers(self, dtype):
...
+        """
+        orig = np.arange(100, dtype=float).reshape((10, 10), order='f')
+        assert not orig.flags.contiguous
+        primary = fits.PrimaryHDU()
+        hdu = fits.CompImageHDU(orig)
+        hdulist = fits.HDUList([primary, hdu])
+        hdulist.writeto(self.temp('test.fits'))
+
+        actual = fits.getdata(self.temp('test.fits'))
+        assert_equal(orig, actual)
```
Comment (pllim) [inline astropy/io/fits/tests/test_image.py]: I wonder if exact match required for floating point comparison will come back to bite us, but let's leave it as-is for now.
```suggestion
        assert_equal(orig, actual)
```

**Comment by saimn** [comment]
@MSeifert04 - The original data is restored after it has been written, and yea that means that a copy is done when the data is noncontiguous, but doing something more efficient is more difficult (#3895).

**Pull request merged**